docs/DeveloperGuide.md

```diff
@@ -1,15 +1,306 @@
+Initially, data validation was being handled by the `Parser` class, however in the principles of avoiding tight coupling 
+and improving cohesion, it was moved back under the `AddModuleCommand` class.
+
+#### 3.4.2 DeleteModuleCommand
+
+##### How the feature is implemented
+The `DeleteModuleCommand` class extends the `Command` class.
+Similar to `AddModuleCommand` class, the constructor `DeleteModuleCommand()` parses the user `input` module code `.toUpperCase()` as the format to fetch an
+instance of `module` from its class. Boolean `successful` field is used to flag successfully added modules in comparison
+to instances where it is not possible to add the `module` as it already exists in the `state`'s `selectedModuleList`.
+It overrides the `execute()` method from the `Command` class, and updates `successful` accordingly, which will later be
+passed on to the overridden `getExecutionMessage()` which displays the result of data validation that the `selectedModule`
+instance is only removed from the `selectedModuleList` if it exists.
+
+##### Why it is implemented this way.
+In order to be able to be able to compare the new instance of `selectedModule` created of the module code the user wants to delete,
+in the constructor against an instance of the module the user has previously added into the `selectedModuleList`, the `equals()`
+method extended from super class `Object` has been overridden to return `true` for instances where `semester` and `module`
+(specifically `moduleCode` attribute from the parent class) are the same, allowing us to validate and remove the desired module.
+
+
+##### Alternatives considered.
+Once again, data validation was being handled by the `Parser` class, however in the principles of avoiding tight coupling
+and improving cohesion, it was moved back under the `DeleteModuleCommand` class.
+
+#### 3.4.3 HelpCommand
+
+#### 3.4.4 SearchModuleCommand
+
+##### How the feature is implemented
+The <code>SearchModuleCommand</code> class extends the <code>Command</code> class.
+It overrides the <code>execute()</code> method from the <code>Command</code> class.
+The <code>execute()</code> method will search for the user input module primarily based on either module code or title,
+with additional parameters of semester and level to narrow down the search results.
+
+##### Why it is implemented this way.
+User may or may not know the exact module code or title. As such, the user can search for the module based on optional 
+parameters such as semester or level. However, the user must input at least the module code or title before additional 
+parameters can be added in order to refine the search.
+
+##### Alternatives considered.
+We thought of implementing the search feature in a way that the required user for multiple inputs and displaying all the
+different results after each input. However, we decided against it as it would be too tedious for the user to input 
+multiple times and the search process will be too long.
```

Comment:
Like how you all got add the alternatives considered, help improve understanding of your point of view and why you choose to do things.